server/src/main/scala/org/http4s/server/staticcontent/FileService.scala

```diff
@@ -58,69 +57,75 @@ object FileService {
         pathPrefix: String = "",
         bufferSize: Int = 50 * 1024,
         cacheStrategy: CacheStrategy[F] = NoopCacheStrategy[F]): Config[F] = {
-      val pathCollector: PathCollector[F] = filesOnly
+      val pathCollector: PathCollector[F] = (f, c, r) =>
+        filesOnly(Path.fromNioPath(f.toPath()), c, r)
       Config(systemPath, pathCollector, pathPrefix, bufferSize, cacheStrategy)
     }
   }
 
   /** Make a new [[org.http4s.HttpRoutes]] that serves static files. */
   private[staticcontent] def apply[F[_]](config: Config[F])(implicit F: Async[F]): HttpRoutes[F] = {
     object BadTraversal extends Exception with NoStackTrace
-    Try(Paths.get(config.systemPath).toRealPath()) match {
-      case Success(rootPath) =>
-        TranslateUri(config.pathPrefix)(Kleisli { request =>
-          def resolvedPath: OptionT[F, Path] = {
-            val segments = request.pathInfo.segments.map(_.decoded(plusIsSpace = true))
-            if (request.pathInfo.isEmpty) OptionT.some(rootPath)
-            else
-              OptionT
-                .liftF(F.catchNonFatal {
-                  segments.foldLeft(rootPath) {
-                    case (_, "" | "." | "..") => throw BadTraversal
-                    case (path, segment) =>
-                      path.resolve(segment)
+    Kleisli
+      .liftF[OptionT[F, *], Any, HttpRoutes[F]] {
+        OptionT.liftF[F, HttpRoutes[F]] {
+          Files[F].realPath(Path(config.systemPath)).attempt.map {
+            case Right(rootPath) =>
+              TranslateUri(config.pathPrefix)(Kleisli { request =>
+                def resolvedPath: OptionT[F, Path] = {
+                  val segments = request.pathInfo.segments.map(_.decoded(plusIsSpace = true))
+                  if (request.pathInfo.isEmpty) OptionT.some(rootPath)
+                  else
+                    OptionT
+                      .liftF(F.catchNonFatal {
+                        segments.foldLeft(rootPath) {
+                          case (_, "" | "." | "..") => throw BadTraversal
+                          case (path, segment) =>
+                            path.resolve(segment)
+                        }
+                      })
+                }
+                resolvedPath
+                  .flatMapF(path =>
+                    F.ifM(Files[F].exists(path, false))(
+                      path.absolute.normalize.some.pure,
```

Comment:
This line is replacing `path.toRealPath(LinkOption.NOFOLLOW_LINKS)`. I'm not 100% that the semantics are exactly the same, but AFAICT making a path into a "real path" _without_ following links is essentially making absolute and then normalizing. This is what the [Javadoc](https://docs.oracle.com/javase/8/docs/api/java/nio/file/Path.html#toRealPath-java.nio.file.LinkOption...-) says for `toRealPath`:
> The precise definition of this method is implementation dependent but in general it derives from this path, an absolute path that locates the same file as this path, but with name elements that represent the actual name of the directories and the file.

Reply:
I took a look at the source for `path.toRealPath(LinkOption.NOFOLLOW_LINKS)` in JDK, at least for Unix:
https://github.com/openjdk/jdk/blob/5ec1cdcaf39229a7d2457313600b0dc2bf8c6453/src/java.base/unix/classes/sun/nio/fs/UnixPath.java#L814

This is what it's doing:
1. `toAbsolutePath()`
2. eliminate "." and also ".." where the previous element is not a link
3. check file exists (without following links)

So, the only part that's different is the `..` handling. _However_, in `FileService` paths that include `..` are already rejected. So, I think it is safe to make this change.